rfc/mediachain-rfc-4.md

```diff
@@ -0,0 +1,330 @@
+participation. At the other end, we want to support curated namespaces
+to combat spam and allow authoritative sources and commercial media
+providers to maintain control of their own data.
+
+In order to implement this model, each participating peer has an
+associated identity corresponding to a public key. The identities
+are intended to be long-lived and eventually have an attached reputation
+score.
+
+All statements published in the system are signed with the key of the
+source of the statement. In order to publish in a moderated namespace,
+the originating peer must present a certificate which allows it to
+publish there. Alternatively, it can convince a peer with the right
+certificate to publish on its behalf. Certificates can be obtained by
+moderators and owners of the namespace. 
+
+This moderation scheme, with delegated administrative control,
+attaches a cost to peer identities because certificates can be
+revoked. This encourages cooperative behavior by individual peers,
+thus avoiding the social cost of cheap identities [1]. At the same
+time, individual peers with publishing rights are free to implement
+their own authentication for their clients.
+
+## Namespace-based Queries and Aggregation
+
+Namespaces provide a convenient way to request information about the same object from several different strata. For example, a query for an image from the NYPL collection and the corresponding OldNYC geocoding data can look like
+
+```
+{nypl:510d47e2-ef28-a3d9-e040-e00a18064a99@nypl,oldnyc}
+```
+
+which would mean "return results for WKI nypl:510d47e2-ef28-a3d9-e040-e00a18064a99 from [nypl](http://digitalcollections.nypl.org/items/510d47e2-ef28-a3d9-e040-e00a18064a99) and [oldnyc](https://www.oldnyc.org/#1557929) namespaces". NYPL can also choose to transclude the statment from the `oldnyc` namespace into `nypl` itself,  "blessing" it.
+
+Other possibilities include filtered aggregation (a `lithographs` namespace that aggregates some objects from `getty`, some from `nypl`, etc), lazy aggregation (logical aggregates that do not publish a stream until requested), etc. Exact semantics TBD.
+
+
+## A Heterogeneous Network of Cooperative Peers
+
+The distributed database is maintained by a heterogeneous peer-to-peer
+network. Peers are operated by different organizations and
+individuals, and contribute different resources to the network.
+
+Each peer maintains some parts of the dataset and has a limited view
+of the mediachain. Peers synchronize their state with peer-to-peer
+interactions, and converge their view of the data structure by
+exchanging published statements.
+
+Statements can be propagated in the network via push, either with
+direct messages or through streaming. For stream propagation, stable
+peers form pubsub overlays for their namespaces and emit published
+statements to their subscribers. Equally well, peers can poll their
+peers for recent updates, allowing them to operate in a disconnected
+fashion.
```

Comment:
I think we want to use a combination of push and pull. 

Always connected peers setup pubsub overlays for online update propagation. We probably want a gossip protocol instead of multicast trees for robustness and simplicity reasons (at the cost of some efficiency).

Less connected clients buffer their updates for pushing when they come back online, at which point they can publish them to the pubsub overlay. For receiving updates they poll other peers who are online.